include/apq/type_traits.h

```diff
@@ -0,0 +1,317 @@
+
+/**
+* Indicates if type is adapted for introspection with Boost.Fusion
+*/
+template <typename T>
+using is_fusion_adapted = std::integral_constant<bool, 
+    ::boost::fusion::traits::is_sequence<T>::value
+>;
+
+/**
+* Indicates if type is adapted for introspection with Boost.Hana
+*/
+template <typename T>
+using is_hana_adapted = std::integral_constant<bool,
+    ::boost::hana::Sequence<T>::value ||
+    ::boost::hana::Struct<T>::value
+>;
+
+/**
+* Indicates if type is a composite. In general we suppose that composite 
+* is a type being adapted for introspection via Boost.Fusion or Boost.Hana,
+* including tuples and compile-time sequences.
+*/
+template <typename T>
+struct is_composite : std::integral_constant<bool,
+    is_hana_adapted<T>::value ||
+    is_fusion_adapted<T>::value
+> {};
+
+
+/**
+* Type traits template forward declaration. Type traits contains information
+* related to it's representation in the database. There are two different
+* kind of traits - built-in types with constant OIDs and custom types with
+* database depent OIDs. The functions below describe neccesary traits.
+* For built-in types traits will be defined there. For custom types user
+* must define traits.
+*/
+template <typename T>
+struct type_traits;
+
+/**
+* Helpers to make size trait constant
+*     bytes - makes fixed size trait
+*     dynamic_size - makes dynamis size trait
+*/
+template <std::size_t n>
+using bytes = std::integral_constant<std::size_t, n>;
+using dynamic_size = void;
+
+/**
+* Helper defines the way for the type traits definitions.
+* Type is undefined then Name type is not defined.
+*     Name - type which can be converted into a string representation
+*         which contain the fully qualified type name in DB
+*     Oid - oid type - provides type's oid in database, may be defined for
+*         built-in types only; custom types have only dynamic oid, depended
+*         from the current state of DB.
+*     Size - size type - provides information about type's object size, may
+*         be specified only if the type's object has fixed size.
+*/
+template <typename Name, typename Oid = null_oid_t, typename Size = dynamic_size>
+struct type_traits_helper {
+    using name = Name;
+    using oid = Oid;
+    using size = Size;
+};
+
+/**
+* By default all non specialized types have unspecified name, name
+* oid and size.
+*/
+template <typename T>
+struct type_traits : type_traits_helper<void> {};
+
+/**
+* Condition indicates if the specified type is built-in for PG
+*/
+template <typename T>
+struct is_built_in : std::integral_constant<bool,
+    !std::is_same_v<typename type_traits<T>::oid, null_oid_t>> {};
+
+template <typename T>
+using is_dynamic_size = std::is_same<typename type_traits<T>::size, dynamic_size>;
+
+/**
+* Function returns type name in Postgre SQL. For custom types user must
+* define this name.
+*/
+template <typename T>
+constexpr auto type_name(const T&) noexcept {
```

Comment:
We need function that returns `boost::hana::string` to support compile time query text generation.

Reply:
If you are talking about query for custom type oids then I think it is better to use this for every type in the oid map:

    typename type_traits<T>::name{};